btcec.go

```diff
@@ -594,29 +609,223 @@ func (curve *KoblitzCurve) Double(x1, y1 *big.Int) (*big.Int, *big.Int) {
+	// in the binary, we want to put a -1 in the lowest bit and get a bunch of
+	// 0s up to the highest bit of consecutive 1s. This is due to this identity:
+	// 2^n + 2^(n-1) + 2^(n-2) + ... + 2^(n-k) = 2^(n+1) - 2^(n-k)
+	// The algorithm thus may need to go 1 more bit than the length of the bits
+	// we actually have, hence bits being 1 bit longer than was necessary.
+	// Since we need to know whether adding will cause a carry, we go from
+	// right-to-left in this addition.
+	var carry, curIsOne, nextIsOne bool
+	// these default to zero
+	retPos := make([]byte, len(k)+1)
+	retNeg := make([]byte, len(k)+1)
+	for i := len(k) - 1; i >= 0; i-- {
+		curByte := k[i]
+		for j := uint(0); j < 8; j++ {
+			curIsOne = curByte&1 == 1
+			if j == 7 {
+				if i == 0 {
+					nextIsOne = false
+				} else {
+					nextIsOne = k[i-1]&1 == 1
+				}
+			} else {
+				nextIsOne = curByte&2 == 2
+			}
+			if carry {
+				if curIsOne {
+					// This bit is 1, so we continue to carry and
+					// don't need to do anything
+				} else {
+					// We've hit a 0 after some number of 1s.
+					if nextIsOne {
+						// We start carrying again since we're starting
+						// a new sequence of 1s.
+						retNeg[i+1] += 1 << j
+					} else {
+						// We stop carrying since 1s have stopped.
+						carry = false
+						retPos[i+1] += 1 << j
+					}
+				}
+			} else if curIsOne {
+				if nextIsOne {
+					// if this is the start of at least 2 consecutive 1's
+					// we want to set the current one to -1 and start carrying
+					retNeg[i+1] += 1 << j
+					carry = true
+				} else {
+					// this is a singleton, not consecutive 1's.
+					retPos[i+1] += 1 << j
+				}
+			}
+			curByte >>= 1
+		}
+	}
+	if carry {
+		retPos[0] = 1
+	}
+
+	return retPos, retNeg
+}
+
 // ScalarMult returns k*(Bx, By) where k is a big endian integer.
 // Part of the elliptic.Curve interface.
 func (curve *KoblitzCurve) ScalarMult(Bx, By *big.Int, k []byte) (*big.Int, *big.Int) {
-	// This uses the left to right binary method for point multiplication:
-
 	// Point Q = ∞ (point at infinity).
 	qx, qy, qz := new(fieldVal), new(fieldVal), new(fieldVal)
 
-	// Point P = the point to multiply the scalar with.
-	px, py := curve.bigAffineToField(Bx, By)
-	pz := new(fieldVal).SetInt(1)
+	// decompose K into k1 and k2 in order to halve the number of EC ops
+	// see Algorithm 3.74 in Guide to Elliptical Curve Cryptography by
+	// Hankerson, et al.
+	k1, k2, signK1, signK2 := curve.splitK(curve.moduloReduce(k))
+
+	// The main equation here to remember is
+	// k * P = k1 * P + k2 * ϕ(P)
+	// P1 below is P in the equation, P2 below is ϕ(P) in the equation
+	p1x, p1y := curve.bigAffineToField(Bx, By)
+	// For NAF, we need the negative point
+	p1yNeg := new(fieldVal).Set(p1y).Negate(1)
+	p1z := new(fieldVal).SetInt(1)
+	// Note ϕ(x,y) = (βx,y), the Jacobian z coordinate is 1, so this math
+	// goes through.
+	p2x := new(fieldVal).Set(p1x).Mul(curve.beta)
```

Comment:
Minor optimization here.  If you use `Mul2` you can multiply the 2 values without having to copy `p1x` first with `Set`.
```Go
p2x := new(fieldVal).Mul(p1x, curve.beta)
```